### PR TITLE
Xyc/liespline

### PR DESCRIPTION
**Pull request by XiangyuChen0212**
change the type of input and output to LieTensor

### CODE REVIEW

File: pypose/module/liesplinese3.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
+import torch
+from torch import nn
+
+class lieSpline(nn.Module):
+
+    def __init__(self):
+        super().__init__()
+        self.B = torch.tensor([[5, 3, -3, 1],
+                               [1, 3, 3, -2],
+                               [0, 0, 0, 1]]) / 6
```
Comment (wang-chen) [inline pypose/module/liesplinese3.py:10]: this will have issues if inputs tensors is on GPU.

File: pypose/module/liesplinese3.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+    def compute_weights(self, time):
+        alpha = torch.arange(4)
```
Comment (wang-chen) [inline pypose/module/liesplinese3.py:13]: this will have issues if inputs tensors is on GPU.

File: pypose/module/liesplinese3.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        tt = time[:, None, :] ** alpha[None, :, None]
+        return self.B @ tt
+
+    def interpolateSE3(self, input_poses, time, inter_begin_and_end = True):
+        """ B spline interpolation in SE3
+        Args:
+            input_poses: input n(n>=4) poses with [batch_size, n, 4, 4] shape
+            time: k interpolation time point with [1, 1, k] shape
+        Returns:
+            output_poses: output (n-3)*k poses with [batch_size, (n-3)*k, 4, 4] shape
```
Comment (wang-chen) [inline pypose/module/liesplinese3.py:23]: Explain `inter_begin_and_end`

Comment (wang-chen) [reply pypose/module/liesplinese3.py:23]: add brief math explanation in docstring

File: examples/module/liespline/liespline.png
Comment (wang-chen) [inline examples/module/liespline/liespline.png:1]: - don't upload generated figures
- follow other examples that generate figures as well, add `.gitignore`, and `save` folders, e.g., [ekf](https://github.com/pypose/pypose/blob/main/examples/module/filter/ekf.py)

File: pypose/module/liesplinese3.py
Comment (wang-chen) [inline pypose/module/liesplinese3.py:1]: Try to generate documentation for this module, you will know how to change the docstring. To do this, refer to this [guideline](https://github.com/pypose/pypose/blob/main/CONTRIBUTING.md#12-writing-documentation)

**Review by wang-chen — CHANGES_REQUESTED** [review]

File: pypose/module/liesplinese3.py (hunk 1 above)
Comment (XiangyuChen0212) [reply pypose/module/liesplinese3.py]: `        alpha = torch.arange(4, dtype=time.dtype, device=time.device)
        tt = time[:, None, :] ** alpha[None, :, None]
        B = torch.tensor([[5, 3, -3, 1],
                               [1, 3, 3, -2],
                               [0, 0, 0, 1]], dtype=time.dtype, device=time.device) / 6
        return B @ tt`

**Comment by XiangyuChen0212** [comment]
@wang-chen Hi, I have modified my code by previous commits.

File: pypose/module/liesplinese3.py (hunk 1 above)
Comment (XiangyuChen0212) [reply pypose/module/liesplinese3.py]: Have modified the code
`
        alpha = torch.arange(4, dtype=time.dtype, device=time.device)
        tt = time[:, None, :] ** alpha[None, :, None]
        B = torch.tensor([[5, 3, -3, 1],
                          [1, 3, 3, -2],
                          [0, 0, 0, 1]], dtype=time.dtype, device=time.device) / 6`

File: pypose/module/liesplinese3.py (hunk 2 above)
Comment (XiangyuChen0212) [reply pypose/module/liesplinese3.py]: Have modified the code 

`alpha = torch.arange(4, dtype=time.dtype, device=time.device)`

File: pypose/module/liesplinese3.py (hunk 3 above)
Comment (XiangyuChen0212) [reply pypose/module/liesplinese3.py]: Have explained the parameter inter_begin_and_end and added math explanation.

File: examples/module/liespline/liespline.png
Comment (XiangyuChen0212) [reply examples/module/liespline/liespline.png:1]: Delete the figure and modified the example.

File: pypose/module/liesplinese3.py
Comment (XiangyuChen0212) [reply pypose/module/liesplinese3.py:1]: Have generated the documentation image

**Review by wang-chen — CHANGES_REQUESTED** [review]
Here are some overall comments:
- No data is stored in the module, thus I suggest defining it as a function instead of a module.
- The name seems not conventional, how about calling it `liesplrep` following [splrep](https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.splrep.html#scipy.interpolate.splrep) in numpy.
- many lines exceeding 90 characters. refer to 2nd point in [developer guideline](https://github.com/pypose/pypose/wiki/Developer-Guideline).
- use `tensor[..., i]` to support batch operation instead of `tensor[:, i]`, refer to 10th point in [developer guideline](https://github.com/pypose/pypose/wiki/Developer-Guideline).
- some display issues in the doc, please view [this pr doc](https://pypose--224.org.readthedocs.build/en/224/generated/pypose.module.LieSpline.html#pypose.module.LieSpline)

**Comment by wang-chen** [comment]
1. Renamed `CSplineR3` to `chspline` (functions name is lower case) and try to work for any batch dim, please test. The point dimsion can also be others, such as 2, or 4, not 3 only.
2. Rename `steps` to `interval`, however, it should have a safe value checking, please add. For example, `interval = 10` will raise errors.

**Comment by wang-chen** [comment]
please add tests, not only examples.

File: pypose/function/spline.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+    K, N = time.shape[-1], data.shape[-2]
+    dargs = {'dtype': time.dtype, 'device': time.device}
+    alpha = torch.arange(4, **dargs).view(-1, 1)
+    tt = time ** alpha
+    B = torch.tensor([[5, 3,-3, 1],
+                      [1, 3, 3,-2],
+                      [0, 0, 0, 1]], **dargs) / 6
+    T_delta = data[..., 0:-3, :].unsqueeze(2).repeat(1, 1, K, 1)
+    w = (B @ tt).unsqueeze(-1)
+    P = torch.stack([data[..., i:i + 4, :] for i in range(N - 3)], dim=1).unsqueeze(-2)
```
Comment (wang-chen) [inline pypose/function/spline.py:192]: please try to remove the loop, you may generate the index first without loops, then use the index for data accessing.

File: pypose/function/spline.py [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+    assert time.shape[0]==time.shape[1]==1, "The time has wrong shape."
+    data = torch.cat((data[..., :1, :], data, data[..., -1:, :]), dim=-2)
+    K, N = time.shape[-1], data.shape[-2]
+    dargs = {'dtype': time.dtype, 'device': time.device}
+    alpha = torch.arange(4, **dargs).view(-1, 1)
+    tt = time ** alpha
+    B = torch.tensor([[5, 3,-3, 1],
+                      [1, 3, 3,-2],
+                      [0, 0, 0, 1]], **dargs) / 6
+    T_delta = data[..., 0:-3, :].unsqueeze(2).repeat(1, 1, K, 1)
```
Comment (wang-chen) [inline pypose/function/spline.py:190]: use `expand` when possible, `repeat` will copy data.

**Review by wang-chen — CHANGES_REQUESTED** [review]
i re-formate some codes, leave some comments for you to address. @XiangyuChen0212

**Review by wang-chen — APPROVED** [review]
